firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/CRCElements.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright 2020, the Pez Globo team and the Pufferfish project contributors
+ *
+ * Chunks.cpp
+ *
+ *  Created on: Nov 3, 2020
+ *      Author: Renji Panicker
+ *
+ * Unit tests to confirm behavior of chunk splitter
+ *
+ */
+
+#include "Pufferfish/Protocols/CRCElements.h"
+
+#include "Pufferfish/HAL/CRCChecker.h"
+#include "Pufferfish/Test/Util.h"
+#include "Pufferfish/Util/Array.h"
+#include "Pufferfish/Util/Endian.h"
+#include "catch2/catch.hpp"
+
+namespace PF = Pufferfish;
+
+SCENARIO("Protocols::CRCElement behaves correctly", "[CRCElement]") {
+  GIVEN("A CRC element") {
+    constexpr size_t buffer_size = 254UL;
+    using TestCRCElementProps = PF::Protocols::CRCElementProps<buffer_size>;
+    using TestCRCElement = PF::Protocols::CRCElement<TestCRCElementProps::PayloadBuffer>;
+    PF::HAL::SoftCRC32 crc32c{PF::HAL::crc32c_params};
+
+    WHEN("when good data is given as input") {
+      auto crc = PF::Util::make_array<uint8_t>(0x98, 0xdb, 0xe3, 0x55);
+      auto body = std::string("\x98\xdb\xe3\x55\x01\x05\x01\x02\x03\x04\x05", 11);
+      auto expected_payload = std::string("\x01\x05\x01\x02\x03\x04\x05", 7);
+      uint32_t expected_crc = 0;
+      PF::Util::read_ntoh(crc.data(), expected_crc);
+
+      TestCRCElementProps::PayloadBuffer payload;
+      TestCRCElement crc_element{payload};
+
+      PF::Util::ByteVector<buffer_size> input_buffer;
+      PF::Util::convertStringToByteVector(body, input_buffer);
+
+      auto parse_status = crc_element.parse(input_buffer);
+      auto crc_check = crc_element.crc();
+      const auto &payload_check = crc_element.payload();
+      THEN("the final status should be ok") {
```

Comment:
I think we're checking more than just the final status here, so the `THEN` should reflect that, e.g. "THEN the body is successfully and correctly parsed". Actually, it could be better if we had three separate `THEN`s.

Reply:
https://github.com/pez-globo/pufferfish-software/blob/feature/firmware-test-protocols/firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/CRCElements.cpp#L206 please review if this what you had in mind. when then/require are added where the function is called, there are some scope issues, i might need to check more on that! will do if this is not what you had in mind.

Reply:
Looks good!